Challenge 23

virginiarcruz opened the pull request:
@fdaciuk  pode me dar uma ajuda?

Estou tentando fazer a operação pelo reduce, só que não tá dando muito certo.
Ao invés de fazer a operação ele concatena os itens do array e se eu fizer um if pelas operações (+, -, x, %), ele só faz a primeira operação.

resultOperation()

Por enquanto estou mostrando o resultado só no console.

virginiarcruz commented:
@fdaciuk ja vi pelo review.
A forma como foi desenvolvida ficou um pouco complicada pra eu entender... mas vou seguir com a modularização do código.

fdaciuk commented on challenge-23/challenge-23.js:
```diff
@@ -31,59 +31,89 @@
+        var oper = item.split('').pop();
+        var removeOp = item.slice(0, -1);
+        var resultOper = inputAll.reduce( function(acumulado, atual, index, array) {
+            acumulado = +removeOp;
```
Oi @virginiarcruz! Desculpe a demora na resposta =)
Mas o maior erro pode estar aqui: procure não modificar variáveis que são passadas por parâmetro nas funções. Sempre que precisar modificar ou computar algum valor, crie uma nova variável =)

Esse valor `acumulado` principalmente, pois ele é o valor que vai vir atualizado a cada iteração no array =)

fdaciuk commented:
Oi @virginiarcruz! O que não ficou muito claro pra você? Me diz que eu vou tentando te explicar com mais detalhes, ok? =)

virginiarcruz commented:
@fdaciuk não entendi muito bem esta parte.

```
var firstValue = acum.slice(0, -1);
var operator = acum.split('').pop();
var lastValue = removeLastItem(actual);
var lastOperator = isLastItemAnOperation(actual) ? actual.split('').pop() : ' ' ;
```
Sendo mais abrangente a função `resultSum() `ficou um pouco confusa a resolução pra mim. Até entendi os passos, mas me perco se for tentar fazer sozinha novamente.

fdaciuk commented:
Oi @virginiarcruz! O que ficou confuso pra você foi como o `reduce` funciona, é isso?

virginiarcruz commented:
@fdaciuk a forma de criação das variáveis e como é feita a captura dos números e operadores.
Pega o primeiro valor com o operador(`firstValue`), depois pega só o operador (`operator`), a partir daí não entendi o `lastValue` e nem o `lastOperator`.

fdaciuk commented:
Ok @virginiarcruz, vamos lá =)

Bom, antes de tudo, é importante entender bem como o `reduce` funciona. Basicamente ele funciona assim:

O `reduce` é um método de array, então ele só funciona com arrays. Ele é uma função que recebe uma outra função por parâmetro.

O que ele faz basicamente é:
- Iterar por cada item do array (como se fosse o `for`);
- A cada iteração ele executa a função que foi passada por parâmetro, e o retorno dessa função ele coloca sempre no valor `accumulated` da próxima iteração.

Exemplificando de forma simples. Imagine esse array:

```js
var arr = [1, 2, 3];
```

Um array simples, com 3 números. Vamos usar o `reduce` para somar todos os números. No final, o `reduce` vai retornar o resultado da soma de todos os valores dessa array:

```js
var result = arr.reduce(function (acc, value) {
  return acc + value;
});

console.log(result); // 6
```

Veja que o resultado final é `6`. O que aconteceu? Basicamente isso:
- Na primeira iteração, o valor utilizado será o primeiro índice do array. Nesse caso, o `1`. Como é a primeira iteração, a variável `acc` não tem valor, correto? Então ela assume o valor do primeiro índice - no caso, o `1` - , e a variável `value` assume o valor do próximo índice - no caso, o `2`.
- Nessa primeira iteração, o retorno da função então vai ser `acc + value`, ou seja: `1 + 2`. Logo, a função vai retornar o número `3`.
- Na próxima iteração, como já foram usados os dois primeiros índices, o `reduce` vai pegar o próximo, no caso o `3`, que vai ser atribuído para a variável `value`.
- Agora, o valor de `acc` vai ser o retorno da função. Então `acc` vai ser igual a `3`.
- Então ele faz novamente o cálculo do `acc + value`, que nesse caso vai ser `3 + 3`, que dá `6`.
- Se o array tivesse mais itens para serem iterados, o `6` seria atribuiído para a variável `acc`.
- Como não tem mais itens no array para iterar, então `6` é o resultado final retornado pelo `reduce`.

E o legal do `reduce` é que, por mais que ele seja um método que só pode ser usado por arrays, ele pode retornar qualquer tipo de valor: números, strings, objetos, outro array. É você quem decide =)

Nesse exemplo eu mostrei como ele pega um array, soma todos os valores e retorna um número no final.

No nosso caso da calculadora, nós pegamos um array de strings, e no final retornamos um número, que é o valor calculado. Funciona dessa forma:

No visor da calculadora, temos o valor: `10x2+3`. Para fazer esse cálculo, nós precisamos separar os números dos operadores, certo? Então a ideia que eu mostrei na resolução era de separar o número e o operador mais próximo à sua direita. No caso, o valor do visor vai se transformar da string:

`'10x2+3'`, em um array: `['10x', '2+', '3']`

Com isso a gente consegue começar a fazer os cálculos.

Como os valores desse array são strings, não tem como calcular antes de converter eles pra número, certo?

Então basicamente a gente precisa pegar só os números antes de fazer o cálculo.

E como fazer isso?

```
var visor =  ['10x', '2+', '3'];
var result = visor.reduce(function (acc, value) {
  // na primeira iteração, `acc` vai ser '10x' e `value` vai ser '2+'. 
  // Precisamos separar o 10, o "x", o 2 e o +
  // separando eles, nós poderemos fazer o primeiro cálculo: `10x2`.
  var firstValue = acc.slice(0, -1); 
  
  // poderíamos fazer também:
  // var firstValue = acc.replace(/\D/g, '');
  // para ele fazer um replace de tudo o que não for número e devolver só o número.
  // fazendo isso, já temos o 10. Precisamos pegar o operador que está entre o 10 e o 2, que é o "x":
  var operator = acc.split('').pop();

  // agora transformamos o "10x" em um array removemos o último 
  // item desse array e passamos esse valor para a variável `operator`.
  // agora precisamos pegar o "2". Ele está no segundo parâmetro, "value", que é igual a "2+".
  // vamoso separar ele do "+":
  var lastValue = value.replace(/\D/g, ''); // vamos deixar assim que fica mais claro, é outra forma de fazer

  // agora já temos o primeiro número, o operador e o segundo número. 
  // mas precisamos do operador do segundo valor, pois no final precisamos 
  // concatenar esse operador no resultado da operação, pois o valor retornado vai ser
  // atribuído na variável `acc`.
  var lastOperator = value.replace(/\d+/g, '');

  // Pronto! Fizemos o `replace` em `value` para substituir tudo o que for número por nada (string vazia).
  // Isso significa que só vai ser passado para a variável `lastOperator` o que NÃO É NÚMERO,
  // no caso, o sinal de "+". Se o valor passado para `value` for o último item do array, ele não
  // vai ter o sinal da operação. No caso, o replace já retorna uma string em branco então, que
  // será concatenada com o resultado.
  // Como é tudo string, precisamos transformar os números em `number` e ver qual é o operador
  // para fazer o cálculo corretamente. Agora o switch:

  switch (operator) {
    // vou colocar só o de multiplicação pra deixar mais simples, mas aqui tem que fazer o case
    // pra todos os operadores que vc pode usar: +, -, * e /
    case 'x': return (Number(firstValue) * Number(lastValue)) + lastOperator;
  }
});
```
Pronto! O retorno da função vai ser primeiro o cálculo do `firstValue` e do `lastValue` (10 * 2), e então concatenamos com o operador que sobrou. Então o retorno da função vai ser: `20+`.

Esse valor `20+` vai ser passado para `acc`. E aí a função é executada novamente, agora com o `acc` tendo como valor `20+` e o `value` o próximo valor do array. Nesse caso, o `3`.

Ela vai fazer todo o processo novamente até que não haja mais itens no array, e com isso temos o resultado do cálculo.

Me diga se ainda ficou alguma coisa complicada de entender, ok? =)

virginiarcruz commented:
@fdaciuk agora entendi direitinho.
Vou tentar fazer novamente a função das operações pra fixar este conteúdo.

Obrigada pela paciência hehehe.

fdaciuk commented:
Legal @virginiarcruz! Tranquilo! Quando precisar, só me marcar com a dúvida! Estou aqui pra te ajudar :D
